Silhoutte plot

neelasha23 opened the pull request:
Add silhouette plot and score.

Commit:
Added silhouette plot

Commit:
Added example and tests

Commit:
Added score

Commit:
lint

Commit:
Added license in test

Commit:
Added doctring

Commit:
Added doctring

Commit:
Minor change

neelasha23 committed:
Merge branch 'master' into silhoutte_plot

Commit:
Renamed function

Commit:
Merge branch 'silhoutte_plot' of https://github.com/neelasha23/sklearn-evaluation into silhoutte_plot

Commit:
Resolve error

Commit:
Removed score

idomic commented:
@neelasha23 tag us when ready for review.
I assume you took the tests module as is? If so try to think if we're missing something (mainly on the integration that didn't exist in the module).
Also, we'll need to add docs, like easy code snippets users can plug and play.

neelasha23 commented:
Yes I'll tag. I did add a test case for the Integration,  but I'll  check if anything more can be added.

Commit:
Changed example

Commit:
added tests

Commit:
Added baseline images

Commit:
Moved examples

Commit:
Deleted example

Commit:
Testcase for params

Commit:
Added example

Commit:
Moved examples

neelasha23 committed:
Delete plot_from_results.py

Commit:
Added mock

Commit:
Merge branch 'silhoutte_plot' of https://github.com/neelasha23/sklearn-evaluation into silhoutte_plot

Commit:
Revert metric change

Commit:
Removed unused import

Commit:
Added examples

neelasha23 committed:
Delete silhouette_plot.py

Commit:
Added testcases

Commit:
Added missing images

neelasha23 commented:
@idomic @edublancas This is ready for review. I have added around 6 test cases apart from the scikit-plot tests, but still the coverage score seems to be stuck at 92%.  Do I need to add anymore code snippets apart from the 4 example scripts?

idomic commented on tests/test_silhouette_plot.py:
```diff
@@ -0,0 +1,144 @@
+        plot.silhouette_plot(X, clf)
+
+
+@patch('sklearn_evaluation.plot.clustering.silhouette_plot_from_results')
```
Please use monkey patch instead of @patch.

idomic reviewed (changes requested):
92% coverage is fine.
Overall looks good, we'd want to add a guide as part of this PR too.
Look here, this can use you as [a sample](https://sklearn-evaluation.readthedocs.io/en/latest/user_guide/compare.html).
We want something similar but for clustering, for users who aren't aware of this we'll add some explanations on when to use it how to use it and what for.

Commit:
Added guide. replaced patch with monkeypatch

neelasha23 commented:
@idomic have added a .ipynb file for silhouette. Do I need to add guide for elbow curve in this PR or create a separate PR?
Haven't added a .rst file, I think it would be having similar content as the `clustering_evaluation.ipynb` file

idomic commented:
@neelasha23 Please include it as part of this PR

idomic commented:
I don't believe there's a way to render the docs locally (I haven't seen it in the readme/tasks).
@edublancas thoughts? Either we push it and check the guide or we add similar scripts to ploomber so devs can test locally

edublancas commented:
there's a way to render docs locally: https://github.com/ploomber/sklearn-evaluation/tree/master/docs

but the instructions might be incomplete. so if you find any issues, let us know so we can help and please document any missing steps

Commit:
Elbow curve testcases

Commit:
Formatting issues

Commit:
Removed img comparison

Commit:
Removed img comparison

Commit:
User guides

neelasha23 commented:
@idomic  @edublancas 
have added test cases and docs for elbow curve as well. A couple of points that I want to highlight:
1. It looks like there's some inconsistency with the matplotlib `@image_comparison` functionality. There were two test cases which passed on my local but failed in the CI so I have removed the image comparison for only those 2 cases. Running the `override_failed_result_images.py` script also didnt help in these cases.  Maybe we should include our own image comparison utility like `yellowbricks` has.
2. While building the docs locally I faced this [issue](https://github.com/ploomber/sklearn-evaluation/issues/71). It might be because sklearn has a module by the same name and some conflict is happening. I renamed the module as `classification_report_skeval.py` and that solved the issue. However I haven't pushed this change as it might be an issue with my local setup. Let me know if I still to push this

idomic commented:
@neelasha23 I see some conflict, please solve it.

idomic commented on src/sklearn_evaluation/plot/clustering.py:
```diff
@@ -141,3 +142,167 @@ def _clone_and_score_clusterer(clf, X, n_clusters):
+    -----------
+
+    X : array-like, shape = [n_samples, n_features]:
+        Data to cluster, where n_samples is the number of samples and
```
Data to cluster -> Cluster data

idomic replied:
Also missing the type within the array

idomic reviewed (commented):
Looks good, please fix comments and conflict.
I still need to run the docs locally to check the guide.

idomic commented:
For 1. we can open an issue about it and move on, I think as long as it works without those 2 tests it's fine.
For 2. I'll build the docs now locally and lets see if this continues.

neelasha23 committed:
Merge branch 'master' into silhoutte_plot

Commit:
Resolved review comments

Commit:
Merge branch 'silhoutte_plot' of https://github.com/neelasha23/sklearn-evaluation into silhoutte_plot

Commit:
Added link

idomic commented:
@yafimvo please review this PR.

yafimvo commented on src/sklearn_evaluation/plot/clustering.py:
```diff
@@ -141,3 +143,168 @@ def _clone_and_score_clusterer(clf, X, n_clusters):
     return clf.fit(X).score(X), time.time() - start
+
+
+def silhouette_plot(X,
```
We need to track the usage of `def silhouette_plot` and `def silhouette_plot_from_results`. We can do it with `@SKLearnEvaluationLogger.log(feature='plot')`

neelasha23 replied:
added

yafimvo commented on tests/test_clustering.py:
```diff
@@ -0,0 +1,209 @@
+    plot.silhouette_plot(X, clf, range_n_clusters=[6], metric='cosine')
+
+
+def test_string_classes():
```
This test generates the file `silhouette_plot_string_classes.png` in the root folder. I suggest saving this file in some dedicated directory for test artifacts that we can add to .gitignore.

neelasha23 replied:
This was a miss from my end. I have removed the saving logic.

yafimvo reviewed (changes requested):
I tested it locally and it looks good. 

2 notes:
1. After running the test `test_string_classes` it generates the file `silhouette_plot_string_classes.png` in the root folder (then generates a new possible commit). I suggest saving this file in some dedicated directory for test artifacts that we can add to .gitignore.
2. We need to track the usage of `def silhouette_plot` and `def silhouette_plot_from_results`. We can do it with `@SKLearnEvaluationLogger.log(feature='plot')`

Commit:
Added log tracking

yafimvo commented:
looking good to me
@idomic @neelasha23

Commit:
Relative import

idomic commented:
Good points, on 1 we should use a fixture to save it into a tmp folder.

idomic commented:
@neelasha23 tag me when this is done and we can merge.

neelasha23 commented:
@idomic  I have fixed @yafimvo 's review comments. On point 1 , it was a miss from my end. I forgot to remove the image save logic, I was just using it for testing purpose. Have removed the logic now. If the docs look okay this is ready

Commit:
Added tolerance to img compare

Commit:
Added version 7

Commit:
Increased tol

Commit:
Increased tol

Commit:
Increased tol

Commit:
Increased tol

Commit:
Increased tol

Commit:
Reverted changes

Commit:
Reverted changes

Commit:
Increased tol, added versions

Commit:
Revert changes

idomic reviewed (approved):
lgtm

idomic commented:
Nice job!

edublancas commented:
the docs are still missing. I see there is a reference to `user_guide/clustering_evaluation` but the notebook isn't there. also, the new plots do not appear on the plots list in the docs.

idomic commented:
He's talking about the [readthedocs](https://sklearn-evaluation.readthedocs.io/en/latest/)

neelasha23 commented:
I'm able to see it when I build the docs in my local:
i think it will be visible if this [PR](https://github.com/ploomber/sklearn-evaluation/pull/85) is merged?
<img width="945" alt="Screenshot 2022-11-29 at 11 32 05 PM" src="https://user-images.githubusercontent.com/29096135/204609730-0bf10556-6e7a-4906-86b4-23beea6a80d2.png">

neelasha23 commented:
I still don't see the guide [here](https://sklearn-evaluation.readthedocs.io/en/latest/)

edublancas commented:
try deleting the cache in your browser: https://sklearn-evaluation.readthedocs.io/en/latest/user_guide/clustering_evaluation.html

neelasha23 commented:
it worked now !
